test/test_train_mp_imagenet.py

```diff
@@ -197,32 +197,39 @@ def train_loop_fn(loader):
       tracker.add(FLAGS.batch_size)
       if lr_scheduler:
         lr_scheduler.step()
-      if x % FLAGS.log_steps == 0:
-        xm.add_step_closure(_train_update, args=(device, x, loss, tracker))
+      if step % FLAGS.log_steps == 0:
+        step = '{}/{}'.format(step, len(train_loader))
+        args = (device, step , loss, tracker, epoch)
+        xm.add_step_closure(_train_update, args=args)
 
-  def test_loop_fn(loader):
-    total_samples = 0
-    correct = 0
+  def test_loop_fn(loader, epoch):
+    total_samples, correct = 0, 0
     model.eval()
-    for data, target in loader:
+    for step, (data, target) in enumerate(loader):
       output = model(data)
       pred = output.max(1, keepdim=True)[1]
       correct += pred.eq(target.view_as(pred)).sum().item()
       total_samples += data.size()[0]
-
+      if step % FLAGS.log_steps == 0:
+        step = '{}/{}'.format(step, len(test_loader))
+        test_utils.print_test_update(device, None, step=step, epoch=epoch)
     accuracy = 100.0 * correct / total_samples
-    test_utils.print_test_update(device, accuracy)
+    test_utils.print_test_update(device, accuracy=accuracy, epoch=epoch)
     return accuracy
 
-  accuracy = 0.0
-  max_accuracy = 0.0
+  accuracy, max_accuracy = 0.0, 0.0
   for epoch in range(1, FLAGS.num_epochs + 1):
+    xm.master_print(
+        'Epoch {} training begin {}'.format(epoch, test_utils.now()))
     para_loader = pl.ParallelLoader(train_loader, [device])
-    train_loop_fn(para_loader.per_device_loader(device))
-    xm.master_print('Finished training epoch {}'.format(epoch))
-
+    train_loop_fn(para_loader.per_device_loader(device), epoch)
+    xm.rendezvous('epoch-train-end')
```

Comment:
Why rendezvous here?

Reply:
I added this is to synchronize the prints, I can remove if we think scrambled prints are ok.